hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,247 @@
+	}
+	qi.dvName = dvName
+	qi.cleanOldCaches(ctx, cl, oldDVs)
+
+	return nil
+}
+
+func (qi *cachedQCOWImage) cleanOldCaches(ctx context.Context, cl client.Client, oldDVs []v1beta1.DataVolume) {
+	logger := ctrl.LoggerFrom(ctx)
+	for _, oldDV := range oldDVs {
+		if oldDV.DeletionTimestamp == nil {
+			logger.Info("deleting an old boot image cache DataVolume", "namespace", oldDV.Namespace, "DataVolume name", oldDV.Name)
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(nodePool)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(nodePool *hyperv1.NodePool) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName:         bootImageDVLabelRoleValue,
+				bootImageDVLabelHosedClusterName: nodePool.Spec.ClusterName,
+			},
+			Annotations: map[string]string{
+				bootImageDVAnnotationHash:          qi.hash,
+				cdiImmediateBindingAnnotation:      "true",
+				cdiDeleteAfterCompletionAnnotation: "false",
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+			Storage: &v1beta1.StorageSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]apiresource.Quantity{
+						corev1.ResourceStorage: volumeSize,
+					},
+				},
+				StorageClassName: storageClass,
+			},
+		},
+	}
+}
+
+func getCacheDVSelector(clusterName string) client.MatchingLabels {
+	return map[string]string{
+		bootImageDVLabelRoleName:         bootImageDVLabelRoleValue,
+		bootImageDVLabelHosedClusterName: clusterName,
```

Comment:
I learned that we should be using the hostedcluster.spec.infraid instead of cluster name to uniquely identify resources like this. I'm doing something similar now to identify the vms, services, and routes we create, https://github.com/openshift/hypershift/pull/2092